src/test/app/Clawback_test.cpp

```diff
@@ -0,0 +1,971 @@
+            env.require(balance(bob, cindy["USD"](-1000)));
+
+            // bob claws back 600 USD from cindy
+            env(claw(bob, cindy["USD"](600)));
+            env.close();
+
+            // cindy has 400 USD left in bob's trustline after clawed by bob
+            env.require(balance(cindy, bob["USD"](400)));
+            env.require(balance(bob, cindy["USD"](-400)));
+
+            // cindy still has 800 USD in alice's trustline
+            env.require(balance(cindy, alice["USD"](800)));
+            env.require(balance(alice, cindy["USD"](-800)));
+        }
+
+        // alice issues USD to both bob and cindy.
+        // when alice claws back from bob, only bob's USD balance is
+        // affected, and cindy's balance remains unchanged, and vice versa.
+        {
+            Env env(*this, features);
+
+            Account alice{"alice"};
+            Account bob{"bob"};
+            Account cindy{"cindy"};
+
+            env.fund(XRP(1000), alice, bob, cindy);
+            env.close();
+
+            auto const USD = alice["USD"];
+
+            // alice sets asfAllowClawback
+            env(fset(alice, asfAllowClawback));
+            env.close();
+            env.require(flags(alice, asfAllowClawback));
+
+            // alice sends 600 USD to bob
+            env.trust(USD(1000), bob);
+            env(pay(alice, bob, USD(600)));
+            env.close();
+
+            env.require(balance(alice, bob["USD"](-600)));
+            env.require(balance(bob, alice["USD"](600)));
+
+            // alice sends 1000 USD to cindy
+            env.trust(USD(1000), cindy);
+            env(pay(alice, cindy, USD(1000)));
+            env.close();
+
+            env.require(balance(alice, cindy["USD"](-1000)));
+            env.require(balance(cindy, alice["USD"](1000)));
+
+            // alice claws back 500 USD from bob
+            env(claw(alice, bob["USD"](500)));
+            env.close();
+
+            // bob's balance is reduced
+            env.require(balance(alice, bob["USD"](-100)));
+            env.require(balance(bob, alice["USD"](100)));
+
+            // cindy's balance is unchanged
+            env.require(balance(alice, cindy["USD"](-1000)));
+            env.require(balance(cindy, alice["USD"](1000)));
+
+            // alice claws back 300 USD from cindy
+            env(claw(alice, cindy["USD"](300)));
+            env.close();
+
+            // bob's balance is unchanged
+            env.require(balance(alice, bob["USD"](-100)));
+            env.require(balance(bob, alice["USD"](100)));
+
+            // cindy's balance is reduced
+            env.require(balance(alice, cindy["USD"](-700)));
+            env.require(balance(cindy, alice["USD"](700)));
+        }
+    }
+
+    void
+    testBidirectionalLine(FeatureBitset features)
+    {
+        testcase("Bidirectional line");
+        using namespace test::jtx;
+
+        // Test when both alice and bob issues USD to each other.
```

Comment:
very interesting test case!
Can we include this scenario also in the specification/documentation. I missed this case.

Reply:
It is written in the spec, under `Clawback` transaction

>Indicates the account which is executing this transaction. The account MUST be the issuer of the asset being clawed back. Note that in the XRP Ledger, trustlines are bidirectional and, under some configurations, both sides can be seen as the "issuer" of an asset. In this specification, the term issuer is used to mean the side of the trustline that has an outstanding balance (i.e. 'owes' the issued asset) that it wishes to claw back.